tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -5773,106 +5756,128 @@ Value *CGMSHLSLRuntime::EmitHLSLInitListExpr(CodeGenFunction &CGF, InitListExpr
+        EltVals, EltQualTys);
     }
   } else if (HLMatrixLower::IsMatrixType(Ty)) {
-    bool isRowMajor = IsRowMajorMatrix(Type, bDefaultRowMajor);
+    DXASSERT(hlsl::IsHLSLMatType(QualTy), "QualType/Type mismatch!");
     // matrix type is struct { vector<Ty, row> [col] };
     // Strip the struct level here.
-    Constant *matVal = C->getAggregateElement((unsigned)0);
-    const RecordType *RT = Type->getAs<RecordType>();
-    RecordDecl *RD = RT->getDecl();
-    QualType EltTy = RD->field_begin()->getType();
-    // When scan, init list scalars is row major.
-    if (isRowMajor) {
-      // Don't change the major for row major value.
-      FlatConstToList(matVal, EltValList, EltTy, Types, bDefaultRowMajor);
-    } else {
-      // Save to tmp list.
-      SmallVector<Constant *, 4> matEltList;
-      FlatConstToList(matVal, matEltList, EltTy, Types, bDefaultRowMajor);
-      unsigned row, col;
-      HLMatrixLower::GetMatrixInfo(Ty, col, row);
-      // Change col major value to row major.
-      for (unsigned r = 0; r < row; r++)
-        for (unsigned c = 0; c < col; c++) {
-          unsigned colMajorIdx = c * row + r;
-          EltValList.emplace_back(matEltList[colMajorIdx]);
+    Constant *RowArrayVal = C->getAggregateElement((unsigned)0);
+    QualType MatEltQualTy = hlsl::GetHLSLMatElementType(QualTy);
+
+    unsigned RowCount, ColCount;
+    hlsl::GetHLSLMatRowColCount(QualTy, RowCount, ColCount);
+
+    // Get all the elements from the array of row vectors.
+    // Matrices are never in memory representation so convert as needed.
+    SmallVector<Constant *, 16> MatElts;
+    for (unsigned r = 0; r < RowCount; ++r) {
+      Constant *RowVec = RowArrayVal->getAggregateElement(r);
+      for (unsigned c = 0; c < ColCount; ++c) {
+        Constant *MatElt = RowVec->getAggregateElement(c);
+        if (MatEltQualTy->isBooleanType()) {
+          DXASSERT(MatElt->getType()->isIntegerTy(1),
+            "Matrix elements should be in their register representation.");
+          MatElt = llvm::ConstantExpr::getZExt(MatElt, Types.ConvertTypeForMem(MatEltQualTy));
         }
+        MatElts.emplace_back(MatElt);
+      }
     }
-  } else if (llvm::ArrayType *AT = dyn_cast<llvm::ArrayType>(Ty)) {
-    QualType EltTy = Type->getAsArrayTypeUnsafe()->getElementType();
-    for (unsigned i = 0; i < AT->getNumElements(); i++) {
-      FlatConstToList(C->getAggregateElement(i), EltValList, EltTy, Types,
-                      bDefaultRowMajor);
-    }
-  } else if (dyn_cast<llvm::StructType>(Ty)) {
-    RecordDecl *RD = Type->getAsStructureType()->getDecl();
-    const CGRecordLayout &RL = Types.getCGRecordLayout(RD);
+
+    // Return the elements in the order respecting the orientation.
+    bool IsRowMajor = hlsl::IsHLSLMatRowMajor(QualTy, bDefaultRowMajor);
+    for (unsigned r = 0; r < RowCount; ++r) {
+      for (unsigned c = 0; c < ColCount; ++c) {
+        unsigned Idx = IsRowMajor ? (r * ColCount + c) : (c * RowCount + r);
```

Comment:
I'm not sure why we are transposing constant matrix based on row/col major.  Since the logical layout is always row major, it should not make any difference for constant initializer, should it?

Reply:
I would feel more comfortable if there was a test that exercised this a couple different ways and made sure the components came out in the right order.

Reply:
I was confused about this too, but the reason it matters is that the constant initializer has to produce a value that will live in memory (where the static variable lives), so it must have its bools as i32s and its matrices in memory packing order such that the loads then properly reconvert to the register representation. Let's see if I can add a test.